pgx/src/datum/sql_entity_graph/sql_graph_entity.rs

```diff
@@ -110,13 +113,40 @@ impl ToSql for SqlGraphEntity {
                 }
             }) {
                 Ok(String::default())
-            } else { item.to_sql(context) },
-            SqlGraphEntity::Type(item) => item.to_sql(context),
+            } else {
+                item.to_sql(context)
+            }
+            SqlGraphEntity::Type(item) if !item.to_sql_config.enabled => Ok(String::default()),
+            SqlGraphEntity::Type(item) => if let Some(callback) = item.to_sql_config.callback {
```

Comment:
I think this decision could be inside `item.to-sql()`? It seems since the config is local to the entity it should decide to have SQL output or not?

Reply:
Yeah I originally lifted the checks into this impl since everything goes through it anyway, and it allows us to avoid creating an extra `SqlGraphEntity` value in the item impl, but thinking on it further, I suspect the optimizer can do well enough on this so as to make it a moot point. I'll move this work into the individual impls.

Reply:
Digging a bit further, there was a little more to my choice here that I forgot, and I'd be interested in your take on what the best approach is.

By delegating all the work to the `ToSql` impl of `SqlGraphEntity`, we can just pass a reference to the current entity to the callback and we're off to the races. If we instead want to delegate the work to the individual `ToSql` impls, things get complicated because we need to wrap the items in the `SqlGraphEntity` enum for genericity (i.e. the ability to handle various entity types with a single callback), but since the `ToSql` impls themselves operate on a reference, we can't get a value of `SqlGraphEntity` without cloning the item, imposing quite a bit of additional overhead.

I think in principle it would make the most sense for the impls to have the responsibility, but given the tradeoffs, leaving control in the hands of the one impl (that of `SqlGraphEntity`) is the better choice. I'm open to a different approach though, just curious what your take is on it.

Reply:
I fiddled around and I understand now.

Thinking on it: If a user decorates say `fn doop()`:

```rust
#[pg_extern(sql = doop_sql)]
fn doop() {}

fn doop_sql(entity: &SqlGraphEntity, pgx: &PgxSql) -> Result<String, Box<dyn std::error::Error + Send + Sync + 'static>> {
  // ...
}
```

Here they **know** entity is a `SqlGraphEntity::Function` variant.  You can see this from your examples:

https://github.com/zombodb/pgx/blob/f5e89cfc0aa88e5451c5da63046c4c328c3a3711/pgx-tests/src/tests/schema_tests.rs#L33-L58

On the other hand, I can totally see the use case of having one of these functions being used by multiple things...

I... think your strategy is good.